data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -0,0 +1,221 @@
+                                 final PluginMetrics pluginMetrics, final ThresholdCheck thresholdCheck, final int retryCount, final long backOffTimeBase) {
+
+        this.cloudWatchLogsClient = cloudWatchLogsClient;
+        this.buffer = buffer;
+        this.logGroup = cloudWatchLogsSinkConfig.getLogGroup();
+        this.logStream = cloudWatchLogsSinkConfig.getLogStream();
+        this.thresholdCheck = thresholdCheck;
+
+        this.retryCount = retryCount;
+        this.backOffTimeBase = backOffTimeBase;
+
+        this.bufferedEventHandles = new ArrayList<>();
+        this.logEventSuccessCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_PUSHED_TO_CWL_SUCCESS);
+        this.requestFailCount = pluginMetrics.counter(REQUESTS_FAILED);
+        this.logEventFailCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_PUSHED_TO_CWL_FAIL);
+        this.requestSuccessCount = pluginMetrics.counter(REQUESTS_SUCCEEDED);
+
+        reentrantLock = new ReentrantLock();
+
+        stopWatch = StopWatch.create();
+        stopWatchOn = false;
+    }
+
+    /**
+     * Function handles the packaging of events into log events before sending a bulk request to CloudWatchLogs.
+     * Implements simple conditional buffer. (Sends once batch size, request size in bytes, or time limit is reached)
+     * @param logs - Collection of Record events which hold log data.
+     */
+    public void output(final Collection<Record<Event>> logs) {
+        reentrantLock.lock();
+
+        try {
+            if (!stopWatchOn) {
+                startStopWatch();
+            }
+
+            for (Record<Event> singleLog: logs) {
+                String logJsonString = singleLog.getData().toJsonString();
+                int logLength = logJsonString.length();
+
+                if (thresholdCheck.isGreaterThanMaxEventSize(logLength + LOG_EVENT_OVERHEAD_SIZE)) {
+                    LOG.warn("Event blocked due to Max Size restriction! {Event Size: " + (logLength + LOG_EVENT_OVERHEAD_SIZE) + " bytes}");
+                    continue;
+                }
+
+                int bufferSizeWithOverHead = (buffer.getBufferSize() + (buffer.getEventCount() * LOG_EVENT_OVERHEAD_SIZE));
+                if ((thresholdCheck.isGreaterThanThresholdReached(getStopWatchTime(),  bufferSizeWithOverHead + logLength + LOG_EVENT_OVERHEAD_SIZE, buffer.getEventCount() + 1) && (buffer.getEventCount() > 0))) {
+                    pushLogs();
+                }
+
+                if (singleLog.getData().getEventHandle() != null) {
+                    bufferedEventHandles.add(singleLog.getData().getEventHandle());
+                }
+                buffer.writeEvent(logJsonString.getBytes());
+            }
+
+            runExitCheck();
+
+        } catch (InterruptedException e) {
+            LOG.error("Caught InterruptedException while attempting to publish logs!");
+            reentrantLock.unlock();
+        }
+    }
+
+    private void pushLogs() throws InterruptedException {
```

Comment:
You could probably split this out into a new class. It could be some retryable log pushing class. This could make it much easier to test.

Reply:
I have extracted a class that deals with this method, it is a little awkward but it does facilitate its readability. It has been extracted into a LogPush class. This should also help if I need to add any more tests. Thank you.